// _organizationprofile/index.js
'use strict';
/******************************************************************************
  DEPENDENCIES = CUSTOM SDK [Custom Software Development Kit]
******************************************************************************/
/******************************************************************************
  PARAMETER = ARGUMENT
******************************************************************************/
// no cli tool
/******************************************************************************
  MODULE INTERNALS & HELPERS
******************************************************************************/
const template      = `<a class="organizationprofile">
<div class="organizationprofile__name"></div>
<div class="organizationprofile__location"></div>
<div class="organizationprofile__logo"></div>
</a>
`
let __              = document.createElement('div');

function organizationprofile (dom, data) { // // @TODO: 'data' maybe also to use for event delegation pattern
  /*var param = { // @TODO: maybe use ES6 or maybe "flow types" to describe INPUT
       dom      : containerDomNode,
       data     : { name: url, logo: url, location: string, website: url }
  }*/
  const COMPONENT   = (__.innerHTML=template,__.childNodes[0]);
  const __name      = COMPONENT.querySelectorAll('.organizationprofile__name')[0];
  const __logo      = COMPONENT.querySelectorAll('.organizationprofile__logo')[0];
  const __location  = COMPONENT.querySelectorAll('.organizationprofile__location')[0];
  const UPDATE      = {
    // @TODO: update functions might be "required"
    // @TODO: update functions might delegate to "child components"
    name              : function UPDATE_name (old, now) {
      __name.innerHTML = now;
    },
    logo              : function UPDATE_logo (old, now) {
      __logo.innerHTML = `<img src="${now}" width="128"/>`;
    },
    location          : function UPDATE_location (old, now) {
      __location.innerHTML = `(${now})`;
    },
    website           : function UPDATE_website (old, now) {
      COMPONENT.setAttribute('href', now);
      COMPONENT.setAttribute('target', '_blank');
    }
  };


  /****** INITIALIZE & DEFAULTS *******/
  // @TODO: use streams to "initialize" and "update"
  UPDATE.name(undefined, data.name || '');
  UPDATE.logo(undefined, data.logo || '');
  UPDATE.location(undefined, data.location || '');
  UPDATE.website(undefined, data.website || '');

  /******** WIRE UP (for SYSTEM UPDATES) ********/
  // @TODO: use streams to "initialize" and "update"
  // data.on('name', UPDATE_name);
  // data.on('logo', UPDATE_logo);
  // data.on('logo', UPDATE_location);

  /******** WIRE UP (for USER UPDATES) ********/
  // __name.addEventListener()
  // __logo.addEventListener()
  // __location.addEventListener()
  // COMPONENT.addEventListener('click', function onClick (event) {
  //   eventstopper(event);
  //   https://github.com/serapath/eventstopper/blob/master/index.js
  //   https://github.com/serapath/webpage
  //   window.open(data.website, '_blank');
  // });

  /******** RETURN *********/
  dom.appendChild(COMPONENT);
  var API = {}; // @TODO: should be an event emitter too - or stream?
  return API;   // @TODO: would this maybe return "a stream" ?
}
/******************************************************************************
  EXPORT
******************************************************************************/
module.exports = organizationprofile;

const css = `:root {
  --Organizationprofile-color1         : #7A68F2;
  --Organizationprofile-color2         : #3627B4;
  --Organizationprofile-color3         : #E2FCEF;
  --Organizationprofile-color4         : #9B287B;
  --Organizationprofile-font1          : Ubuntu Mono, Open Sans, monaco, sans-serif;
  --Organizationprofile-font2          : Ubuntu Mono, Open Sans, monaco, sans-serif;
  --Organizationprofile-font-size1     : 6.5vw;
  --Organizationprofile-font-size2     : 3vw;
  --Organizationprofile-font-size3     : 2.3vw;
  --Organizationprofile-borderwidth    : 4vw;

  --Organizationprofile-mini-font-size1      : 6.7vw;
  --Organizationprofile-mini-font-size2      : 5.5vw;
  --Organizationprofile-mini-font-size3      : 5vw;
  --Organizationprofile-mini-borderwidth     : 6vw;
  --Organizationprofile-mini-margin          : 0vw;

  --Organizationprofile-watch-font-size1      : 6.7vw;
  --Organizationprofile-watch-font-size2      : 5.5vw;
  --Organizationprofile-watch-font-size3      : 5vw;
  --Organizationprofile-watch-borderwidth     : 6vw;
  --Organizationprofile-watch-margin          : 0vw;

  --Organizationprofile-smartphone-font-size1   : 7.5vw;
  --Organizationprofile-smartphone-font-size2   : 4.2vw;
  --Organizationprofile-smartphone-font-size3   : 3.6vw;
  --Organizationprofile-smartphone-borderwidth  : 4vw;
  --Organizationprofile-smartphone-margin       : 0vw;

  --Organizationprofile-tablet-font-size1   : 6vw;
  --Organizationprofile-tablet-font-size2   : 4vw;
  --Organizationprofile-tablet-font-size3   : 3vw;
  --Organizationprofile-tablet-borderwidth  : 6vw;
  --Organizationprofile-tablet-margin       : 8vw;

  --Organizationprofile-desktop-font-size1   : 4. 5vw;
  --Organizationprofile-desktop-font-size2   : 3vw;
  --Organizationprofile-desktop-font-size3   : 2.5vw;
  --Organizationprofile-desktop-borderwidth  : 5vw;
  --Organizationprofile-desktop-margin       : 16vw;

  --Organizationprofile-highres-font-size1   : 4vw;
  --Organizationprofile-highres-font-size2   : 2.5vw;
  --Organizationprofile-highres-font-size3   : 2vw;
  --Organizationprofile-highres-borderwidth  : 4vw;
  --Organizationprofile-highres-margin       : 20vw;
}
.organizationprofile            {
  display                       : flex;
  flex-direction                : column;
  text-align                    : center;
  padding-left                  : 30px;
  padding-right                 : 30px;
  margin-left                   : 30px;
  border                        : 5px solid var(--Organizationprofile-color2);
  margin                        : 5px;
  text-decoration               : none;
  background-color              : var(--Organizationprofile-color1);
}
.organizationprofile__name      {
  color                         : var(--Organizationprofile-color4);
  font-family                   : var(--Organizationprofile-font2);
  font-size                     : var(--Organizationprofile-font-size1);
  text-align                    : center;
  font-weight                   : 900;
}
.organizationprofile__location  {
  color                         : var(--Organizationprofile-color3);
  font-family                   : var(--Organizationprofile-font2);
  font-size                     : var(--Organizationprofile-font-size2);
  text-align                    : center;
  font-weight                   : 900;
}
@media all and (min-width: 0px) and (max-width: 321px) {
  /* mini */
  .organizationprofile__name      {
    font-size                     : var(--Organizationprofile-mini-font-size1);
  }
  .organizationprofile__location  {
    font-size                     : var(--Organizationprofile-mini-font-size2);
  }
}
@media all and (min-width: 321px) and (max-width: 641px) {
  /* watch */
  .organizationprofile__name      {
    font-size                     : var(--Organizationprofile-watch-font-size1);
  }
  .organizationprofile__location  {
    font-size                     : var(--Organizationprofile-watch-font-size2);
  }
}

@media all and (min-width: 641px) and (max-width: 769px) {
  /* smartphone */
  .organizationprofile__name      {
    font-size                     : var(--Organizationprofile-smartphone-font-size1);
  }
  .organizationprofile__location  {
    font-size                     : var(--Organizationprofile-smartphone-font-size2);
  }
}

@media all and (min-width: 769px) and (max-width: 1025px) {
  /* tablet */
  .organizationprofile__name      {
    font-size                     : var(--Organizationprofile-tablet-font-size1);
  }
  .organizationprofile__location  {
    font-size                     : var(--Organizationprofile-tablet-font-size2);
  }
}

@media all and (min-width: 1025px) and (max-width: 1201px) {
  /* desktop */
  .organizationprofile__name      {
    font-size                     : var(--Organizationprofile-desktop-font-size1);
  }
  .organizationprofile__location  {
    font-size                     : var(--Organizationprofile-desktop-font-size2);
  }
}

@media all and (min-width: 1201px) {
  /* highres */
  .organizationprofile__name      {
    font-size                     : var(--Organizationprofile-highres-font-size1);
  }
  .organizationprofile__location  {
    font-size                     : var(--Organizationprofile-highres-font-size2);
  }
}
`
